pkg/mads/reconcile/snapshotter_test.go

```diff
@@ -0,0 +1,275 @@
+							Name: "backend-01",
+							Mesh: "default",
+						},
+						Spec: mesh_proto.Dataplane{
+							Networking: &mesh_proto.Dataplane_Networking{
+								Inbound: []*mesh_proto.Dataplane_Networking_Inbound{{
+									Interface: "192.168.0.1:80:8080",
+									Tags: map[string]string{
+										"service": "backend",
+									},
+								}},
+							},
+						},
+					},
+				},
+				expected: mads_cache.NewSnapshot("", nil),
+			}),
+			Entry("Mesh with Prometheus enabled and some Dataplanes", testCase{
+				meshes: []*mesh_core.MeshResource{
+					{
+						Meta: &test_model.ResourceMeta{
+							Name: "default",
+							Mesh: "default",
+						},
+					},
+					{
+						Meta: &test_model.ResourceMeta{
+							Name: "demo",
+							Mesh: "demo",
+						},
+						Spec: mesh_proto.Mesh{
+							Metrics: &mesh_proto.Metrics{
+								Prometheus: &mesh_proto.Metrics_Prometheus{
+									Port: 1234,
+									Path: "/non-standard-path",
+								},
+							},
+						},
+					},
+				},
+				dataplanes: []*mesh_core.DataplaneResource{
+					{
+						Meta: &test_model.ResourceMeta{
+							Name: "backend-01",
+							Mesh: "default",
+						},
+						Spec: mesh_proto.Dataplane{
+							Networking: &mesh_proto.Dataplane_Networking{
+								Inbound: []*mesh_proto.Dataplane_Networking_Inbound{{
+									Interface: "192.168.0.1:80:8080",
+									Tags: map[string]string{
+										"service": "backend",
+										"env":     "prod",
+									},
+								}},
+							},
+						},
+					},
+					{
+						Meta: &test_model.ResourceMeta{
+							Name: "backend-02",
+							Mesh: "demo",
+						},
+						Spec: mesh_proto.Dataplane{
+							Networking: &mesh_proto.Dataplane_Networking{
+								Inbound: []*mesh_proto.Dataplane_Networking_Inbound{{
+									Interface: "192.168.0.2:443:8443",
+									Tags: map[string]string{
+										"service": "backend",
+										"env":     "intg",
+									},
+								}},
+							},
+						},
+					},
+					{
+						Meta: &test_model.ResourceMeta{
+							Name: "web-01",
+							Mesh: "demo",
+						},
+						Spec: mesh_proto.Dataplane{
+							Networking: &mesh_proto.Dataplane_Networking{
+								Inbound: []*mesh_proto.Dataplane_Networking_Inbound{{
+									Interface: "192.168.0.3:3000:3000",
+									Tags: map[string]string{
+										"service": "web",
+										"env":     "test",
+									},
+								}},
+							},
+							Metrics: &mesh_proto.Metrics{
+								Prometheus: &mesh_proto.Metrics_Prometheus{
+									Port: 8765,
+									Path: "/even-more-non-standard-path",
+								},
+							},
+						},
+					},
+				},
+				expected: mads_cache.NewSnapshot("", map[string]envoy_cache.Resource{
+					"/meshes/demo/dataplanes/backend-02": &observability_proto.MonitoringAssignment{
+						Name: "/meshes/demo/dataplanes/backend-02",
+						Targets: []*observability_proto.MonitoringAssignment_Target{{
+							Labels: map[string]string{
+								"__address__": "192.168.0.2:1234",
```

Comment:
What is the use case of multiple targets?

shouldn't metrics path be in target labels also? In proto definition the docs says
```
  // Describes a single target that needs to be monitored.
  message Target {

    // Labels associated with that particular target.
    //
    // E.g.,
    // `[
    //    "__address__" :      "192.168.0.1:8080",
    //    "__metrics_path__" : "/metrics"]`,
    //    "instance" :         "backend-01",
    //  ]`.
    map<string, string> labels = 1;
  }
```

Reply:
This is done to meet constraints of Prometheus' `file_sd` implementation and [custom-sd](https://github.com/prometheus/prometheus/tree/master/documentation/examples/custom-sd) adapter that is recommended for use by all `file_sd`-based `sd`s.

More detailed version:

* `Prometheus` model for all `sd`s except for `file_sd` looks like this:
   ```
   // Group is a set of targets with a common label set(production , test, staging etc.).
   type Group struct {
	// Targets is a list of targets identified by a label set. Each target is
	// uniquely identifiable in the group by its address label.
	Targets []model.LabelSet
	// Labels is a set of labels that is common across all targets in the group.
	Labels model.LabelSet

	// Source is an identifier that describes a group of targets.
	Source string
   }
   ```
* that is why Kuma's `MonitoringAssignment` was designed to be close that model
* however, `file_sd` uses different model for reading data from a file
   ```
   struct {
	Targets []string       `yaml:"targets"`
	Labels  model.LabelSet `yaml:"labels"`
   }
   ```
   Notice that `Targets` is just a list of addresses rather than a list of `model.LabelSet`
* because of that mismatch, some form of conversion is unavoidable inside `kuma-prometheus-sd`
* the next component that imposes its constraints is [custom-sd](https://github.com/prometheus/prometheus/tree/master/documentation/examples/custom-sd) - adapter that is recommended for use by all `file_sd`-based `sd`s
* this adapter is doing conversion from `Prometheus` model into `file_sd` model and it expects that `Targets` has only 1 label - `__address__` and the rest of the labels come from `Labels`
* so, we need to convert `MonitoringAssignment` into a model that [custom-sd](https://github.com/prometheus/prometheus/tree/master/documentation/examples/custom-sd) expects. It could happen on server side, it could happen on client side. Given that we're trying to minimize amount of logic on the client side, the choice was made in favour of server side

Reply:
Thanks for in-depth explanation. Can you leave the comment in the code with shorter explanation that target should only contain address because it's converted to file_sd eventually?

Reply:
Done